### ISSUE TITLE
Single relationship difference between versions 5.2.8 and 5.3.3  [DATAGRAPH-1329]

### ISSUE THREAD

**Issue by spring-issuemaster** [spring-projects/spring-data-neo4j#1890]
**[ivangfr](https://jira.spring.io/secure/ViewProfile.jspa?name=JIRAUSER49850)** opened **[DATAGRAPH-1329](https://jira.spring.io/browse/DATAGRAPH-1329?redirect=false)** and commented

Hi,

I've migrated a project from **spring boot** version **2.2.8** (that have by default **spring-data-neo4j** version **5.2.8**) to **2.3.3** (whose **spring-data-neo4j** version is **5.3.3**). However, I am facing a weird issue in the latest version.

The full source code can be found here (https://github.com/ivangfr/springboot-neo4j-caffeine).

I will try to keep short. So, some code will be removed.

I have 2 classes: **City** and **Restaurant**

```java
@NodeEntity
public class City {

    @Id
    @GeneratedValue(strategy = UuidStrategy.class)
    private String id;

    private String name;

    @Relationship(type = "LOCATED_IN", direction = Relationship.INCOMING)
    private Set<Restaurant> restaurants = new LinkedHashSet<>();

}

@NodeEntity
public class Restaurant {

    @Id
    @GeneratedValue(strategy = UuidStrategy.class)
    private String id;

    private String name;

    @Relationship(type = "LOCATED_IN")
    private City city;

    ...

}
```

I have an endpoint where I can change the city of the restaurant (https://github.com/ivangfr/springboot-neo4j-caffeine/blob/master/restaurant-api/src/main/java/com/mycompany/restaurantapi/rest/RestaurantController.java#L71)

So, let's say that I have 2 cities, **Berlin** and **Rome**, and one restaurant called **Happy_Pizza** that is located in **Berlin**.

When I updated the location of the restaurant **Happy_Pizza** from **Berlin** to **Rome**, I have the following results:
- **spring-data-neo4j** version **5.2.8**: (/)Correct! The connection between **Happy_Pizza-Berlin** is removed and a new connection **Happy_Pizza-Rome** is created;
- **spring-data-neo4j** version **5.3.3**: (x)Error! The restaurant stays in both cities, i.e, the connection between **Happy_Pizza-Berlin** is NOT removed and a new connection **Happy_Pizza-Rome** is created;

You can use this test case (https://github.com/ivangfr/springboot-neo4j-caffeine/blob/master/restaurant-api/src/test/java/com/mycompany/restaurantapi/RestaurantIntegrationTest.java#L86) to reproduce the issue.

Thanks!

 

 

 

 

---
No further details from [DATAGRAPH-1329](https://jira.spring.io/browse/DATAGRAPH-1329?redirect=false)

**Comment by michael-simons** [comment]
Hi @ivangfr. I hope this is you (the one who originally opened the JIRA Ticket). If not, sorry and please ignore this. 

So, the explanation here is perfectly valid and 100% correct.
Both Spring Data Neo4j versions (5.2.x and 5.3.x) use the same Neo4j-OGM version (3.2.x) underneath, so there's no change.

However, there's one big change in Spring Boot 2.3, that got initiated by us:

https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.3-Release-Notes#neo4j

> The open session in view interceptor for Neo4j is now disabled by default. If you need to enable it again, use the standard spring.data.neo4j.open-in-view property.

Find the reasoning behind it here: https://info.michael-simons.eu/2020/02/03/spring-data-neo4j-neo4j-ogm-and-osiv/

In short: It's better for the server load in most all cases, but can cause issues as this here.

The correct solution is to define well known transactional boundaries. A dedicated restaurant service  is perfect. I would not put a `@Transactional` on the controller level, but one layer below and define a service that orchestrates the repository calls. 

The underlying OGM needs that transaction as it opens a dedicated session for each transaction. In that session there's a context caching the relationships. If those are multiple transactions (`findBy...` being the first, `save` the second), than those contexts will not know of each other and therefor exhibit the behavior you see.
